mscore/importmxmlpass1.cpp

```diff
@@ -2473,6 +2482,231 @@ void MusicXMLParserPass1::setFirstInstr(const QString& id, const Fraction stime)
             }
       }
 
+//---------------------------------------------------------
+//   notations
+//---------------------------------------------------------
+
+/**
+ Parse the /score-partwise/part/measure/note/notations node.
+ */
+
+void MusicXMLParserPass1::notations(MxmlStartStop& tupletStartStop)
+      {
+      Q_ASSERT(_e.isStartElement() && _e.name() == "notations");
+      //_logger->logDebugTrace("MusicXMLParserPass1::note", &_e);
+
+      while (_e.readNextStartElement()) {
+            if (_e.name() == "tuplet") {
+                  QString tupletType       = _e.attributes().value("type").toString();
+
+                  // ignore possible children (currently not supported)
+                  _e.skipCurrentElement();
+
+                  if (tupletType == "start")
+                        tupletStartStop = MxmlStartStop::START;
+                  else if (tupletType == "stop")
+                        tupletStartStop = MxmlStartStop::STOP;
+                  else if (tupletType != "" && tupletType != "start" && tupletType != "stop") {
+                        _logger->logError(QString("unknown tuplet type '%1'").arg(tupletType), &_e);
+                        }
+                  }
+            else {
+                  _e.skipCurrentElement();        // skip but don't log
+                  }
+            }
+
+      Q_ASSERT(_e.isEndElement() && _e.name() == "notations");
+      }
+
+//---------------------------------------------------------
+//   determineTupletFractionAndFullDuration
+//---------------------------------------------------------
+
+/**
+ Split duration into two factors where fullDuration is note sized
+ (i.e. the denominator is a power of 2), 1/2 < fraction <= 1/1
+ and fraction * fullDuration equals duration.
+ */
+
+void determineTupletFractionAndFullDuration(const Fraction duration, Fraction& fraction, Fraction& fullDuration)
+      {
+      fraction = duration;
+      fullDuration = Fraction(1, 1);
+      // move denominator's powers of 2 from fraction to fullDuration
+      while (fraction.denominator() % 2 == 0) {
+            fraction *= 2;
+            fraction.reduce();
+            fullDuration *= Fraction(1, 2);
+            }
+      // move numerator's powers of 2 from fraction to fullDuration
+      while ( fraction.numerator() % 2 == 0) {
+            fraction *= Fraction(1, 2);
+            fraction.reduce();
+            fullDuration *= 2;
+            fullDuration.reduce();
+            }
+      // make sure 1/2 < fraction <= 1/1
+      while (fraction <= Fraction(1, 2)) {
```

Comment:
Not sure whether it is applicable here, but the case of `fraction > 1` is not handled here so it can end up being really more than 1. For example, if it happens so that `duration == 3/4` (some 4:3 tuplet?), then on this line we will have
```
fraction == 3/1;
fullDuration == 1/4;
```
so the function will yield the values that do not match what is probably intended here.